src/Spago/Command/Build.purs

```diff
@@ -140,42 +96,193 @@ run opts = do
+    unless (Map.isEmpty duplicateModules) do
+      die $
+        ( Dodo.lines
+            [ Dodo.text $ "Detected " <> show (Map.size duplicateModules) <> " modules with the same module name across 2 or more packages defined in this workspace."
+            , duplicateModules
+                # Map.toUnfoldable
+                # Array.mapWithIndex
+                    ( \idx (Tuple m ps) ->
+                        Dodo.lines
+                          [ Dodo.text $ show (idx + 1) <> ") Module \"" <> m <> "\" was defined in the following packages:"
+                          , Dodo.indent
+                              $ Dodo.lines
+                              $ map
+                                  ( \(Tuple pkg moduleFilePath) ->
+                                      Dodo.text $ "- " <> PackageName.print pkg <> "   at path: " <> moduleFilePath
+                                  )
+                              $ Array.sort ps
+                          ]
+                    )
+                # Dodo.lines
+            ] :: Docc
+        )
 
 -- TODO: if we are building with all the packages (i.e. selected = Nothing),
 -- then we can use the graph to remove outdated modules from `output`!
 
 type BuildGlobsOptions =
   { withTests :: Boolean
   , depsOnly :: Boolean
-  , selected :: Array WorkspacePackage
-  , dependencies :: Map PackageName Package
+  , selected :: WorkspacePackage
+  , dependencies :: PackageMap
   }
 
 getBuildGlobs :: BuildGlobsOptions -> Set FilePath
 getBuildGlobs { selected, dependencies, withTests, depsOnly } =
   Set.fromFoldable $ projectGlobs <> monorepoPkgGlobs <> dependencyGlobs <> [ BuildInfo.buildInfoPath ]
   where
   -- Here we select the right globs for a monorepo setup with a bunch of packages
-  projectGlobs = join case depsOnly of
+  projectGlobs = case depsOnly of
     true -> []
     false ->
       -- We just select all the workspace package globs, because it's (1) intuitive and (2) backwards compatible
-      map workspacePackageGlob selected
+      workspacePackageGlob selected
+
+  testGlobs = case withTests of
+    true -> WithTestGlobs
+    false -> NoTestGlobs
+
+  workspacePackageGlob :: WorkspacePackage -> Array String
+  workspacePackageGlob p = Config.sourceGlob testGlobs p.package.name (WorkspacePackage p)
+
+  { yes: monorepoPkgs, no: dependencyPkgs } = partition isWorkspacePackage $ Map.toUnfoldable dependencies
+
+  -- depsOnly means "no packages from the monorepo", so we filter out the workspace packages
+  dependencyGlobs = (Tuple.uncurry $ Config.sourceGlob NoTestGlobs) =<< dependencyPkgs
+  monorepoPkgGlobs
+    | depsOnly = []
+    | otherwise = (Tuple.uncurry $ Config.sourceGlob NoTestGlobs) =<< monorepoPkgs
+
+type BuildWorkspaceGlobsOptions =
+  { withTests :: Boolean
+  , depsOnly :: Boolean
+  , workspacePackages :: Array WorkspacePackage
+  , dependencies :: Map PackageName Package
+  }
+
+getEntireWorkspaceGlobs :: BuildWorkspaceGlobsOptions -> Set FilePath
```

Comment:
Are these implementations different enough to warrant a split? It feels like we could cover everything with a few datatypes (I mean, the previous implementation was covering everything, so why not) - adding a different one means they can eventually diverge, which I'd rather not do